### PR TITLE
Add inter-grid transfer operators

### PR DESCRIPTION
**Pull request by peterrum**

### CODE REVIEW

File: doc/news/changes/minor/20200529MunchPrieto
Comment (kronbichler) [inline doc/news/changes/minor/20200529MunchPrieto:4]:
```suggestion
(Peter Munch, Laura Prieto Saavedra, 2020/05/29)
```

File: include/deal.II/multigrid/mg_transfer_interface.h
Comment (tjhei) [inline include/deal.II/multigrid/mg_transfer_interface.h:45]:
```suggestion
 * @note Normally, this data structure does not need to be filled by users, but, one
```

File: include/deal.II/multigrid/mg_transfer_interface.h
Comment (tjhei) [inline include/deal.II/multigrid/mg_transfer_interface.h:46]:
```suggestion
 *   can use the utility functions provided in the namespace
```

File: include/deal.II/multigrid/mg_transfer_interface.h
Comment (tjhei) [inline include/deal.II/multigrid/mg_transfer_interface.h:60]:
```suggestion
  unsigned int n_dofs_coarse_cell;
```
?

File: include/deal.II/multigrid/mg_transfer_interface.h
Comment (tjhei) [inline include/deal.II/multigrid/mg_transfer_interface.h:55]:
```suggestion
  unsigned int n_coarse_cells;
```

File: include/deal.II/multigrid/mg_transfer_interface.h
Comment (tjhei) [inline include/deal.II/multigrid/mg_transfer_interface.h:65]:
```suggestion
  unsigned int n_dofs_fine_cell;
```

File: include/deal.II/multigrid/mg_transfer_interface.h
Comment (tjhei) [inline include/deal.II/multigrid/mg_transfer_interface.h:32]: we typically spell out namespaces, can you use ``Tools`` or ``Utilities``?

File: include/deal.II/multigrid/mg_transfer_interface.h
Comment (tjhei) [inline include/deal.II/multigrid/mg_transfer_interface.h:118]: Most of our multigrid related classes start with ``MG...``. How does this class relate to ``MGTransfer*``?

File: include/deal.II/multigrid/mg_transfer_interface.h
Comment (tjhei) [inline include/deal.II/multigrid/mg_transfer_interface.h:132]: why is this not templated on vector type?

File: include/deal.II/multigrid/mg_transfer_interface.h
Comment (tjhei) [inline include/deal.II/multigrid/mg_transfer_interface.h:195]: The naming is not ideal. Can it be a replacement for the existing MGTransferMatrixFree? We can think about backwards-compatibility separately.

File: include/deal.II/multigrid/mg_transfer_interface.h
Comment (tjhei) [inline include/deal.II/multigrid/mg_transfer_interface.h:211]: this needs documentation.

File: include/deal.II/multigrid/mg_transfer_interface.h
Comment (tjhei) [inline include/deal.II/multigrid/mg_transfer_interface.h:269]: triangulation or dofhandler? Or is this more generic? Are these vectors the same size?

This is probably unrelated to multigrid and could go somewhere else (vectortools?).

File: include/deal.II/multigrid/mg_transfer_interface.h
Comment (tjhei) [inline include/deal.II/multigrid/mg_transfer_interface.h:358]: asserts to check number of levels etc. fit

File: include/deal.II/multigrid/mg_transfer_interface.h
Comment (tjhei) [inline include/deal.II/multigrid/mg_transfer_interface.h:369]: why 0, could we use invalid_unsigned_int?

File: include/deal.II/multigrid/mg_transfer_interface.h
Comment (tjhei) [inline include/deal.II/multigrid/mg_transfer_interface.h:399]: why do you only need to touch the finest vector?

File: include/deal.II/multigrid/mg_transfer_interface.templates.h
Comment (tjhei) [inline include/deal.II/multigrid/mg_transfer_interface.templates.h:180]: use more () when using ``?``

File: include/deal.II/multigrid/mg_transfer_interface.h
Comment (peterrum) [reply include/deal.II/multigrid/mg_transfer_interface.h:399]: Why should you fill the vectors on the levels? They are filled during restriction, aren't they?

File: include/deal.II/multigrid/mg_transfer_interface.h
Comment (peterrum) [reply include/deal.II/multigrid/mg_transfer_interface.h:269]: > This is probably unrelated to multigrid and could go somewhere else (vectortools?).

But for now I would like to keep in this PR since it uses the same infrastructure as the global-coarsening algorithm is using.

File: include/deal.II/multigrid/mg_transfer_interface.h
Comment (peterrum) [reply include/deal.II/multigrid/mg_transfer_interface.h:195]: As it is now, it is not a replacement yet. I have not incorporated the interface matrices since I did not need them for global coarsening and in the hp context. 

Nevertheless, merging both would be nice. For example the following code snippet:

https://github.com/dealii/dealii/blob/9e97cd69e4d72012f9d913c8f0cae69e8f8042a4/source/multigrid/mg_transfer_matrix_free.cc#L373-L508

is more or less identically in my code (with the difference that different local prolongation matrices can be used depending on the cells and that cells might have children or not).

File: include/deal.II/multigrid/mg_transfer_interface.h
Comment (peterrum) [reply include/deal.II/multigrid/mg_transfer_interface.h:118]: What about `MGTwoLevelTransfer`? It is the transfer operator between two multigrid leveles (however they are defined).

File: include/deal.II/multigrid/mg_transfer_interface.h
Comment (peterrum) [reply include/deal.II/multigrid/mg_transfer_interface.h:32]: Done: `MGTransferUtilities`.

**Comment by peterrum** [comment]
@tjhei Sorry for the late reply! The idea is following. 

Let's consider for now two levels (the level with more dofs I will call fine level; the other one is the coarse level). The class `MGTwoLevelTransfer` is responsible for transfer between these two levels. However, within levels we might need to distinguish how to perform element-wise prolongate/restrict: 
- hp-context: restriction/prolongation between element of different polynomial degrees
- global coarsening: restriction/prolongation to children or application of identity matrix (if coarse cell is not refined)

This information is encrypted in the class `MGTransferScheme`. `MGTwoLevelTransfer` practically loops over all `MGTransferScheme`s.

The class `MGTransferMatrixFreeNew` (or however it will be called) is a collection of `MGTwoLevelTransfer`s. 

In the namespace `MGTransferUtilities`, I have collected functions that set up `MGTwoLevelTransfer` for hp-context and global coarsening.

File: include/deal.II/multigrid/mg_transfer_interface.h
Comment (tjhei) [reply include/deal.II/multigrid/mg_transfer_interface.h:195]: maybe MGTransferGlobalCoarsening?

**Comment by tjhei** [comment]
Maybe move VectorRepartitioner into separate PR?
What about custom constraints?

File: tests/multigrid-transfer/mg_transfer_a_01.cc
Comment (peterrum) [reply tests/multigrid-transfer/mg_transfer_a_01.cc]: Indeed, we don't need it. I have removed it.

**Comment by peterrum** [comment]
> maybe MGTransferGlobalCoarsening?

Done!

> Maybe move VectorRepartitioner into separate PR?

I have removed it. Once this PR is merged, I will open the PR for that.

> What about custom constraints?

Could you have a look at the utility functions. There users have the possibility to pass constraint matrices. In what way are these different from *custom constraints*?

File: include/deal.II/multigrid/mg_transfer_interface.templates.h
Comment (tjhei) [inline include/deal.II/multigrid/mg_transfer_interface.templates.h:45]: what about degree=0?

File: include/deal.II/multigrid/mg_transfer_interface_util.h
Comment (tjhei) [inline include/deal.II/multigrid/mg_transfer_interface_util.h:37]: is this just FE_Q or more generic?

File: include/deal.II/multigrid/mg_transfer_interface.h
Comment (peterrum) [reply include/deal.II/multigrid/mg_transfer_interface.h:97]: Nope. It is the local index so that the vector can be accessed directly without any indirections.

File: include/deal.II/multigrid/mg_transfer_interface.h
Comment (peterrum) [reply include/deal.II/multigrid/mg_transfer_interface.h:102]: Same here.

File: include/deal.II/multigrid/mg_transfer_interface_util.h
Comment (peterrum) [reply include/deal.II/multigrid/mg_transfer_interface_util.h:37]: Currently `FE_Q` and `FE_DGQ` (and I guess `FE_DGQArbitraryNodes`). I guess it should be the same as in the case `MGTransferMatrixFree`.

File: include/deal.II/multigrid/mg_transfer_interface.templates.h
Comment (peterrum) [reply include/deal.II/multigrid/mg_transfer_interface.templates.h:45]: We can think about `degree=0`. But let's leave that for a follow-up PR.

**Comment by peterrum** [comment]
@tjhei I have made the changes.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.h
Comment (tjhei) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.h:202]: this documentation is outdated. Please mention p and h global coarsening.

File: include/deal.II/multigrid/mg_transfer_global_coarsening_util.h
Comment (tjhei) [inline include/deal.II/multigrid/mg_transfer_global_coarsening_util.h:27]: Please document the namespace. Also, the name is not ideal, is it?

File: include/deal.II/multigrid/mg_transfer_global_coarsening_util.h
Comment (tjhei) [inline include/deal.II/multigrid/mg_transfer_global_coarsening_util.h:37]: Based on the documentation, I have no idea what this function does. Would this be more appropriate as a static member of MGTransferTwoLevel? I assume that this is what the function reports on.

File: include/deal.II/multigrid/mg_transfer_global_coarsening_util.h
Comment (tjhei) [inline include/deal.II/multigrid/mg_transfer_global_coarsening_util.h:46]: Make this a constructor of MGTwoLevelTransfer?

File: include/deal.II/multigrid/mg_transfer_global_coarsening_util.h
Comment (tjhei) [inline include/deal.II/multigrid/mg_transfer_global_coarsening_util.h:63]: Same.

File: include/deal.II/multigrid/mg_transfer_global_coarsening_util.h
Comment (tjhei) [inline include/deal.II/multigrid/mg_transfer_global_coarsening_util.h:71]: With all three functions gone, we could get rid of this namespace and file.

**Comment by peterrum** [comment]
@tjhei I have following changes:
- remove utility namspace and add the functions to `MGTwoLevelTransfer` (with new names `reinit_geometric_transfer()` and `reinit_polynomial_transfer()`)
- remove old files
- remove the template argument `MeshType` (since `hp::DoFHandler` is now part of `DoFHandler`)

Le's not over-think the structure of the internal functions. My guess is once we want to support additional vector types, we can reuse many functions.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.h
Comment (tjhei) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.h:137]: does this need to be public?

File: tests/multigrid-transfer/global_id_01.cc
Comment (tjhei) [inline tests/multigrid-transfer/global_id_01.cc:17]: update

File: tests/multigrid-transfer/mg_transfer_a_01.cc
Comment (tjhei) [inline tests/multigrid-transfer/mg_transfer_a_01.cc:18]: "Test MGTwoLevelTransfer with mesh coarsening"

File: tests/multigrid-transfer/mg_transfer_a_01.cc
Comment (tjhei) [inline tests/multigrid-transfer/mg_transfer_a_01.cc:152]: The test output seems to be really large (github doesn't even show it). Can you see if this can be reduced somewhat?

File: tests/multigrid-transfer/mg_transfer_p_01.cc
Comment (tjhei) [inline tests/multigrid-transfer/mg_transfer_p_01.cc:18]: Test MGTwoLevelTransfer with polynomial coarsening

File: tests/multigrid-transfer/mg_transfer_p_01.cc
Comment (tjhei) [inline tests/multigrid-transfer/mg_transfer_p_01.cc:144]: same here, the files seem to be quite big?

File: tests/multigrid-transfer/mg_transfer_p_02.cc
Comment (tjhei) [inline tests/multigrid-transfer/mg_transfer_p_02.cc:18]: same here, add a one line description

File: tests/multigrid-transfer/mg_transfer_support_01.output
Comment (tjhei) [inline tests/multigrid-transfer/mg_transfer_support_01.output:10]: beautiful!

File: tests/multigrid-transfer/test.h
Comment (tjhei) [inline tests/multigrid-transfer/test.h:1]: I just realized that this is a new folder. Don't you need to add a CMakeList.txt? Also, maybe we should name this folder multigrid-global-coarsening ?

**Review by tjhei — COMMENTED** [review]
Thanks for the quicj edits. Things look much better now. I have more complaints though, sorry.

**Comment by peterrum** [comment]
@tjhei I have addressed your comments. Once you are happy, I will squash.

**Comment by tjhei** [comment]
Looks good, please squash.

**Comment by peterrum** [comment]
@tjhei Done!

**Review by tjhei — APPROVED** [review]
This looks good to me, but I will let somebody else to take a look. @kronbichler ?

File: include/deal.II/multigrid/mg_transfer_global_coarsening.h
Comment (kronbichler) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.h:166]:
```suggestion
     * Flag if the finite element on the fine cells are continuous. If yes,
```

File: include/deal.II/multigrid/mg_transfer_global_coarsening.h
Comment (kronbichler) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.h:148]: I am not sure about the name. Do you mean
```suggestion
    unsigned int dofs_per_cell_coarse;
```
(and similarly for `n_dofs_fine_cell`). The reason why I'm asking is that I think of a global quantity with `n_dofs`, whereas this seems local.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.h
Comment (kronbichler) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.h:167]:
```suggestion
     * the multiplicity of DoF sharing a vertex/line as wellas constraints have to be taken in
```

File: include/deal.II/multigrid/mg_transfer_global_coarsening.h
Comment (kronbichler) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.h:183]:
```suggestion
     * DoF indices of the coarse cells, expressed in indices local to the MPI rank.
```

File: include/deal.II/multigrid/mg_transfer_global_coarsening.h
Comment (kronbichler) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.h:265]: I am not sure I understand what `l` refers to. I guess it is the global number of levels. Maybe skip, same in the line below?
```suggestion
   * Constructor taking an operator for each level (minimum requirement is
```

File: include/deal.II/multigrid/mg_transfer_global_coarsening.h
Comment (kronbichler) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.h:113]: Since this does not adhere to the general interface `MGTransferBase`, do you think the `to_level` integer in `prolongate` is still needed? Below I see that this gets set to `invalid_unsigned_int`, which suggests to me that a two-level transfer should not need the index.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.h
Comment (kronbichler) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.h:421]: Why do you need `update_ghost_values` here? To me, ghost values should only be held close to the place where we need to read remote vector entries; otherwise all global vector operations like `Vector::add()` call `update_ghost_values` which is likely not intended.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h
Comment (kronbichler) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:60]: I know that other cases are supposed to happen later; my question would be if that follow-up code would then go to a generic version with `-1` template argument (or `0` in the tensor product evaluators) so we do not bail out.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h
Comment (kronbichler) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:318]: Is there a reason to use the trailing underscores in the variable names? We otherwise don't use that convention in deal.II.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h
Comment (kronbichler) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:1092]: This is a candidate for #9544.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h
Comment (kronbichler) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:1653]: This is the place I was referring to in my comment above about the default case `degree=-1`.

**Review by kronbichler — APPROVED** [review]
Apart from a few comments, this looks good to me.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h
Comment (peterrum) [reply include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:60]: I see your point. Let's leave it to a follow-up PR.

**Comment by peterrum** [comment]
@kronbichler  I have addressed your comments!

**Review by kronbichler — APPROVED** [review]

**Pull request merged**
